File: src/main/java/com/test/processor/work/ReceiverB.java
package com.test.processor.work;

import com.rabbitmq.client.*;
import com.test.util.ConnectionUtil;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

/**
 * @author 费世程
 * @date 2020/4/7 16:18
 */
public class ReceiverB {

  private static final String QUEUQ_NAME = "work_queue_test";

  public static void main(String[] args) {
    try {
      Connection connection = ConnectionUtil.Companion.getConnection();
      final Channel channel = connection.createChannel();
      //设置每个消费者同时只能处理一条消息（能者多劳），在手动ack下才生效
      channel.basicQos(1);
      channel.queueDeclare(QUEUQ_NAME,false,false,false,null);
      DefaultConsumer consumer=new DefaultConsumer(channel){
        @Override
        public void handleDelivery(String consumerTag,
                                   Envelope envelope,
                                   AMQP.BasicProperties properties,
                                   byte[] body) throws IOException {
          String msg = new String(body, "UTF-8");
          System.out.println("消费者2 received :" + msg);
          channel.basicAck(envelope.getDeliveryTag(),false);
          try {
            TimeUnit.SECONDS.sleep(2L);
          } catch (InterruptedException e) {
            e.printStackTrace();
          }
        }

      };
      channel.basicConsume(QUEUQ_NAME,false,consumer);

    } catch (Exception e) {
      System.err.println("消费者2--接收消息出现异常：" + e.getMessage());
    }

  }
}

File: src/main/java/com/test/processor/subscribe/direct/Receiver_B.kt
package com.test.processor.subscribe.direct

import com.rabbitmq.client.QueueingConsumer
import com.test.processor.subscribe.fanout.Subscribe_receiver_a
import com.test.util.ConnectionUtil

/**
 * @author 费世程
 * @date 2019/10/24 14:23
 */
class Receiver_B {

  companion object{
    private val QUEUE_NAME="test_queue_direct_2"
    private val EXCHANGE_NAME="test_exchange_direct"

    fun receive(){
      val connection=ConnectionUtil.getConnection()
      val channel=connection.createChannel()

      channel.queueDeclare(QUEUE_NAME,false,false,false,null)

      //绑定队列到交换机
      channel.queueBind(QUEUE_NAME,EXCHANGE_NAME,"insert")

      channel.basicQos(1)

      val consumer=QueueingConsumer(channel)
      //监听队列，手动返回完成
      channel.basicConsume(QUEUE_NAME,false,consumer)

      //获取消息
      while(true){
        val delivery=consumer.nextDelivery()
        val message=String(delivery.body)
        println("[receiver-b] : $message")
        Thread.sleep(10)
        channel.basicAck(delivery.envelope.deliveryTag,false)
      }
    }
  }

}